Forbid macro att on function rules

radumereuta opened the pull request:
Fixes: #1852

I'm assuming all of these attributes are not allowed on function rules: macro|macro-rec|alias|alias-rec
How about claims. What attributes are not allowed?

radumereuta committed:
Forbid macro att on function rules

Fixes: #1852

radumereuta commented:
This is failing: https://github.com/kframework/k/blob/master/k-distribution/tests/regression-new/proof-tests/deposit/spec/deposit-symbolic.k#L8
Are we sure we want to enforce this restriction?

radumereuta committed:
Fix tests.

dwightguth commented:
Claims can't be macros so you can forbid all of those attributes on claims.

dwightguth commented on kernel/src/main/java/org/kframework/compile/checks/CheckFunctions.java:
```diff
@@ -83,4 +85,26 @@ public void apply(KApply k) {
+        new RewriteAwareVisitor(true, errors) {
+            @Override
+            public void apply(KApply k) {
+                if (isRHS() || k.klabel() instanceof KVariable || !m.attributesFor().contains(k.klabel())) {
```
Instead of checking isRHS, which will return true outside of any rewrite, you probably want to check `isRHS() && !isLHS()`

dwightguth reviewed (changes requested):
Please add a test that this check correctly handles rules with function context. You don't seem to have implemented this the usual way, so I'd like to be certain.

radumereuta committed:
Expand tests

radumereuta commented:
@dwightguth I added more tests.
I also added the check to `context` and `context alias`. Please confirm which attributes are allowed because I couldn't find hard evidence of what is allowed where.

radumereuta committed:
Update test to avoid duplicate rules

rv-jenkins committed:
Merge branch 'master' into fix1852-funcRuleAttErr
